src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -401,17 +377,46 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             return new VersionedSemanticRange(result, csharpResponse.ResultId, csharpResponse.IsFinalizedCSharp);
         }
 
+        // Internal for testing only
+        internal static bool TryGetMinimalCSharpRange(RazorCodeDocument codeDocument, [NotNullWhen(true)] out Range? range)
+        {
+            var csharpDoc = codeDocument.GetCSharpDocument();
+
+            // If there aren't any source mappings, there's no C# code in the Razor doc.
+            if (csharpDoc.SourceMappings.Count == 0)
+            {
+                range = null;
+                return false;
+            }
+
+            // We only need to colorize the portions of the generated doc that correspond with a C# mapping.
+            // To accomplish this while minimizing the amount of work we need to do, we'll only colorize the
+            // range spanning from the first C# span to the last C# span.
+            var csharpSourceText = codeDocument.GetCSharpSourceText();
+            var start = csharpSourceText.Lines.GetLinePosition(csharpDoc.SourceMappings[0].GeneratedSpan.AbsoluteIndex);
```

Comment:
I don't think source mappings are guaranteed to be in-order in relation to the C# document's ordering. For instance, this:

```Razor
@functions {
    // Content!!
}

@DateTime.Now
```

Outputs this:
![image](https://user-images.githubusercontent.com/2008729/142916150-ac64de78-fa00-43d9-ae12-64dc8f2c6e37.png)

Note the @DateTime.Now exists prior to the `@functions` content

Reply:
Ah, this is what I was afraid of. I thought I was just being paranoid though lol.